DR-1852 Add DRS support for signing files in Azure

myessail opened the pull request.

Commit:
initial pass

nmalfroy commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -205,12 +221,51 @@ public DRSAccessURL getAccessUrlForObjectId(
 
+    if (wrapper.isGcp()) {
+      return signGoogleUrl(snapshot, matchingAccessMethod);
+    } else {
```
To be safe, I'd do:
```suggestion
    } else if (wrapper.isAzure()) {
```

nmalfroy commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -196,7 +208,11 @@ public DRSAccessURL getAccessUrlForObjectId(
+    BillingProfileModel billingProfileModel =
+        profileService.getProfileById(snapshot.getProfileId(), authUser);
+
+    CloudPlatformWrapper wrapper = CloudPlatformWrapper.of(billingProfileModel.getCloudPlatform());
```
nit, but I've been naming this variable platform:
```suggestion
    CloudPlatformWrapper platform = CloudPlatformWrapper.of(billingProfileModel.getCloudPlatform());
```

nmalfroy commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -196,7 +208,11 @@ public DRSAccessURL getAccessUrlForObjectId(
-    GoogleProjectResource projectResource = snapshot.getProjectResource();
+
+    BillingProfileModel billingProfileModel =
+        profileService.getProfileById(snapshot.getProfileId(), authUser);
```
I know we talked about this but actually, you already have this information with the snapshot and this saves another round trip to the DB:
```suggestion
        snapshot.getFirstSnapshotSource().getDataset().getDatasetSummary().getDefaultBillingProfile();
```

nmalfroy commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -205,12 +221,51 @@ public DRSAccessURL getAccessUrlForObjectId(
+    return object.getAccessMethods().stream()
+        .filter(
+            drsAccessMethod ->
+                drsAccessMethod.getType().equals(TypeEnum.GS)
```
I don't think you want this to be required to be a `TypeEnum.GS`

nmalfroy commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -205,12 +221,51 @@ public DRSAccessURL getAccessUrlForObjectId(
+              resourceConfiguration.getAppToken(UUID.fromString(profileModel.getTenantId())),
+              blobUrl.getBlobContainerName());
+
+      return new DRSAccessURL().url(sourceClientFactory.createReadOnlySasUrlForBlob(blobUrl.getBlobName()));
```
Probably worth adding a parameter to the `createReadOnlySasUrlForBlob` method to be able to specify a timeout

giventocode replied:
`BlobCrl` now has the ability to set a time out and create SAS tokens with the required permissions.

nmalfroy commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -205,12 +221,51 @@ public DRSAccessURL getAccessUrlForObjectId(
+      fsObject = fileService.lookupSnapshotFSItem(
+          snapshotService.retrieveAvailableSnapshotProject(snapshot.getId()), drsId.getFsObjectId(), 1);
+
+      BlobUrlParts blobUrl = BlobUrlParts.parse(fsObject.getPath());
```
Not for this PR, but I wonder if we may want to store in the object store

nmalfroy reviewed (changes requested):
Great start! A couple of comments.  Also, I think you're going to want to not return the access method as a GS type

Commit:
access id stuff

Commit:
Merge branch 'develop' into my-azure-url

Commit:
changes

Commit:
Merge branch 'develop' into my-azure-url

Commit:
ugly

Commit:
test

samanehsan commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDao.java:
```diff
@@ -580,12 +580,12 @@ FireStoreDirectoryEntry computeDirectory(
-    Collections.sort(crc32cCollection);
-    String crc32cConcat = StringUtils.join(crc32cCollection, StringUtils.EMPTY);
-    String crc32cChecksum = fireStoreUtils.computeCrc32c(crc32cConcat);
+    //    Collections.sort(crc32cCollection);
```
Was this leftover from debugging?

tlangs commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -199,21 +217,78 @@ public DRSAccessURL getAccessUrlForObjectId(
+      DRSAccessMethod matchingAccessMethod =
+          getAccessMethodMatchingAccessId(accessId, drsObject, TypeEnum.HTTPS)
+              .orElseThrow(
+                  () ->
```
This could be stored as a `Supplier<IllegalArgumentException>` so that it isn't repeated

pshapiro4broad replied:
I think a static method would also work, and would probably read better than a local variable. That is, a method that accepts `accessId` and `objectId`.
```
  static void noMatchingError(String accessId, String objectId) throws IllegalArgumentException() {
    throw new IllegalArgumentException(...);
  }
```

tlangs commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDao.java:
```diff
@@ -580,12 +580,12 @@ FireStoreDirectoryEntry computeDirectory(
 
     // Update the directory in place
-    dirEntry.checksumCrc32c(crc32cChecksum).checksumMd5(md5Checksum).size(totalSize);
+    dirEntry.checksumCrc32c("").checksumMd5(md5Checksum).size(totalSize);
```
Why empty string?

tlangs commented on src/test/java/bio/terra/service/filedata/DrsServiceTest.java:
```diff
@@ -87,9 +112,10 @@ public void before() throws Exception {
             .description("description")
             .path("file.txt")
-            .gspath("gs://path/to/file.txt")
+            .gspath("https://core.windows.net/blahblah")
```
That doesn't look like a gspath to me. Perhaps we rename the field `cloudPath`?

Commit:
fix test i broke

Commit:
gspath rename

Commit:
supplier

Commit:
Merge branch 'develop' into my-azure-url

Commit:
update from new methods

tlangs commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -74,6 +81,15 @@
+  private final AzureContainerPdao azureContainerPdao;
+  private final AzureBlobStorePdao azureBlobStorePdao;
+
+  private Supplier<IllegalArgumentException> illegalArgumentExceptionSupplier =
```
This can be declared at the method level instead of the class level.

tlangs commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -199,21 +223,66 @@ public DRSAccessURL getAccessUrlForObjectId(
+              .orElseThrow(illegalArgumentExceptionSupplier);
+      return signGoogleUrl(snapshot, matchingAccessMethod);
+    } else if (wrapper.isAzure()) {
+      DRSAccessMethod matchingAccessMethod =
```
it looks like `matchingAccessMethod` isn't used.

Commit:
review comments/spotbugs

samanehsan commented on src/test/java/bio/terra/service/filedata/DrsServiceTest.java:
```diff
@@ -70,28 +92,46 @@ public void before() throws Exception {
+            .bucketResourceId(bucketResourceId.toString());
     when(fileService.lookupSnapshotFSItem(snapshotProject, drsId.getFsObjectId(), 1))
-        .thenReturn(fsFile);
+        .thenReturn(azureFsFile);
```
You can remove this line because you already have this mocked on line 178. Also, I think it would get overridden by line 133 anyways since that takes the same inputs but returns the `googleFsFile`

nmalfroy replied:
I'm going to make it mock a google and azure object separately

samanehsan commented on src/main/java/bio/terra/service/filedata/azure/blobstore/AzureBlobStorePdao.java:
```diff
@@ -176,6 +180,29 @@ public boolean deleteFile(FireStoreFile fireStoreFile) {
+    String blobName = blobParts.getBlobName();
+    BlobSasPermission permission = new BlobSasPermission().setReadPermission(true);
+    BlobSasTokenOptions blobSasTokenOptions =
+        new BlobSasTokenOptions(Duration.ofMinutes(15), permission, "");
```
I'm not sure how the contentDisposition parameter gets used, but I've seen other places in the code use the name of whatever class calls it: https://github.com/DataBiosphere/jade-data-repo/blob/develop/src/main/java/bio/terra/service/filedata/azure/util/BlobContainerCopier.java#L204

snf2ye replied:
Also from that same reference, maybe moving the duration into a variable would be good? Looks like the copier uses a 24 hour timeout?

nmalfroy replied:
Agree it should be a variable.  Also, @saman, the content disposition is the field we can use to arbitrarily encode values.  The plan is to encode the user.  That leads me to think this variable should probably take a Authed user object

samanehsan commented on src/test/java/bio/terra/service/filedata/DrsServiceTest.java:
```diff
@@ -102,8 +142,8 @@ public void before() throws Exception {
-    assertThat(drsObject.getSize(), is(fsFile.getSize()));
-    assertThat(drsObject.getName(), is(fsFile.getPath()));
+    assertThat(drsObject.getSize(), is(googleFsFile.getSize()));
+    assertThat(drsObject.getName(), is(googleFsFile.getPath()));
```
The size and path are the same as the `azureFsFile` Maybe make those different to ensure that you're still getting the right one?

samanehsan commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -199,21 +218,70 @@ public DRSAccessURL getAccessUrlForObjectId(
+              .orElseThrow(illegalArgumentExceptionSupplier);
+      return signGoogleUrl(snapshot, matchingAccessMethod);
+    } else if (wrapper.isAzure()) {
+      getAccessMethodMatchingAccessId(accessId, drsObject, TypeEnum.HTTPS)
```
Just for my own understanding, why do we need to get the access method for Azure when it doesn't get used?

nmalfroy replied:
Is this just to validate that this is a good access id?

nmalfroy commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -199,21 +218,70 @@ public DRSAccessURL getAccessUrlForObjectId(
+    if (wrapper.isGcp()) {
+      DRSAccessMethod matchingAccessMethod =
+          getAccessMethodMatchingAccessId(accessId, drsObject, TypeEnum.GS)
+              .orElseThrow(illegalArgumentExceptionSupplier);
```
I think stylistically, we've always just done this inline instead of defining an explicit supplier variable

nmalfroy commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -199,21 +218,70 @@ public DRSAccessURL getAccessUrlForObjectId(
+                snapshotService.retrieveAvailableSnapshotProject(snapshot.getId()),
+                drsId.getFsObjectId(),
+                1);
+        return signAzureUrl(billingProfileModel, fsItem);
```
I think you could make `signAzureUrl` take a cloudPath rather than fsItem, in which case, you could actually use the result of  `getAccessMethodMatchingAccessId` on line 241 (to avoid multiple calls)

nmalfroy replied:
Err...is the reason we're not so we can get the storage account resource?

nmalfroy replied:
Eh, that might be a bigger change. TODO, we should do some catching on storage account lookups but I think we're ok for now

nmalfroy commented on src/main/java/bio/terra/service/filedata/FSFileInfo.java:
```diff
@@ -33,12 +33,12 @@ public FSFileInfo createdDate(String createdDate) {
-  public String getGspath() {
-    return gspath;
+  public String getCloudPath() {
+    return cloudPath;
```
this is cool but I just want to make sure that this continues to work with existing data in firestore.  I think we left that model with gsPath but I just want to make sure we validate when we merge and deploy on dev

nmalfroy commented on src/test/resources/dataset-minimal-sample.json:
```diff
@@ -1,5 +1,5 @@
-{"id":"sample3","participant_id":"participant_1","date_collected":"2019-02-27"}
-{"id":"sample4","participant_id":"participant_2","date_collected":"2019-02-27"}
-{"id":"sample5","participant_id":"participant_2","date_collected":"2019-02-27"}
+{"id":"sample1","participant_id":"participant_1","date_collected":"2019-02-27", "file": "8b28cf94-d624-4ce2-bb7f-a5cf122943fc"}
```
I think these changes are left over from manual tesitng

nmalfroy reviewed (changes requested):
Nice progress.  A couple of comments

nmalfroy committed:
Address PR comments

- fixup drs service unit test
- clean up test data file
- pass in user's email and signature location to the pdao layer

nmalfroy committed:
Merge branch 'develop' into my-azure-url

nmalfroy committed:
Fix compilation failure

nmalfroy committed:
Merge branch 'develop' into my-azure-url

# Conflicts:
#	src/main/java/bio/terra/service/filedata/azure/blobstore/AzureBlobStorePdao.java
